### PR TITLE
Add Dockerfile and usange examples

### PR DESCRIPTION
**Pull request by simonmysun**
What I did:

- Adding `lxml` in `./server/requirements.txt`
- Adding `npm:@rollup/wasm-node` override in `./client/package.json`
- Adding `Dockerfile` for building Docker image and example usages

Because of the compilation of `lxml`, etc, the building time of this docker image is quite long. I suggest publishing ready-built image to [Docker Hub](https://docs.docker.com/guides/walkthroughs/publish-your-image/) or [Github Docker Registory](https://docs.github.com/en/actions/publishing-packages/publishing-docker-images).

### CODE REVIEW

**Review by Crissium — COMMENTED** [review]
Could you please tell me why `lxml` is necessary? As far as I can remember it's used for:
- XDXF conversion when the module `xdxf2html` isn't available
- Extracting text from HTML articles when building full text search index
- Parsing the release Atom in the auto updater

By the way, should we remove the auto updater in the docker version? This way we have one fewer dependency (`requests`).

Edit: but indeed for the Docker version we'd better make it as complete as possible. I don't think it should be a non-optional requirement though (for ordinary users who have no need for full-text search and could use the module `xdxf2html`).

File: Dockerfile
Comment (Crissium) [inline Dockerfile:1]: Xapian (used for full text search) is in the official repo of Alpine Linux. Should we install it as well?

File: docker-compose.yml
Comment (Crissium) [inline docker-compose.yml:1]: Actually there's another one that had better be mapped: `~/.cache/SilverDict`.

**Comment by simonmysun** [comment]
Thank you for the review. 

My work so far:

- I have reverted 419f7c1. I didn't dig into details. I simply encountered errors and found an import without try catch. I thought it was intentional. Sorry. 
- I added some arguments for building Docker image to allow adding `lxml`, enabling full text search, etc. 

However I don't know how to test the optional features. 

- `lxml` can be built but I don't know if it is used
- `sibel` can be installed and I have put `de.aff` and `de.dic` under `~/.silverdict/hunspell`, but when I restart the application it says `No spellchecking library found`. `hunspell` fails to install and I haven't yet figured out why.
- `xapian` can be installed but when I try full text search it returns 500.
- `opencc` can be installed but I don't see where I can use it. 

Could you please guild me through? You can build image with my branch with `docker build --tag silverdict https://github.com/simonmysun/SilverDict.git`.

Regarding mouting of `~/.cache/SilverDict`, I don't think it is necessary. The cache is generated fast enough and will remain in the container. If you deploy the application as a service, the regeneration only happens when the container is removed, i.g. docker image is upgraded. But we can make it optional for users, though.

Regarding removal of auto updater, I believe it would be ideal but does not worth efforts.

**Comment by Crissium** [comment]
> * `lxml` can be built but I don't know if it is used

It is only used in conjunction with xapian to build better full text search indexes. So I believe we had better enable it when the user chooses to enable xapian.

> * `sibel` can be installed and I have put `de.aff` and `de.dic` under `~/.silverdict/hunspell`, but when I restart the application it says `No spellchecking library found`. `hunspell` fails to install and I haven't yet figured out why.

That's pretty strange. It shouldn't happen. I'll later build the image and test myself.

> * `xapian` can be installed but when I try full text search it returns 500.

I think that's expected behaviour. A full text search index has to be present (it could be built in Settings) to make it work.

> * `opencc` can be installed but I don't see where I can use it.

If you add the language zh to a dictionary group, which includes let's say dictionaries in both the simplified and the traditional scripts, then it would return 门/門 when either is inputted.

And thanks for your wonderful advice regarding the best practices of docker!

**Comment by Crissium** [comment]
Sorry I wasn't exactly clear in the last comment. When I built the image, I used this Docker compose file:
```yaml
version: '3.8'
services:
  silverdict:
    build: 
      context: .
      dockerfile: Dockerfile
      args:
        ENABLE_FULL_TEXT_SEARCH: 'true'
        ENABLE_MORPHOLOGY_ANALYSIS: 'true'
        ENABLE_CHINESE_CONVERSION: 'true'
    volumes:
      - /home/ellis/Workspace/SilverDict/Docker/Testing/data/:/home/silverdict/.silverdict/
      # - /path/to/cache/:/home/silverdict/.cache/SilverDict/ # Optional mounting of cache
    restart: unless-stopped
    ports:
      - 2628:2628
```

I set the args to true following a guide I found, but in your Readme it was suggested that they should be set to `1`. If a variable is `1`, would the condition `"$ENABLE_MORPHOLOGY_ANALYSIS" = "true"` work as expected? Sorry for my silly questions but I don't really have much experience with Docker.

**Comment by simonmysun** [comment]
Sorry I'm currently away from keyboard.

In Dockerfile I used [environment replacement](
https://docs.docker.com/reference/dockerfile/#environment-replacement)
syntax like `${ENABLE_MORPHOLOGY_ANALYSIS:+libhunspell}` which will result
in libhunspell when `$ENABLE_MORPHOLOGY_ANALYSIS` is set, otherwise empty.
This should work when you set them to 'true'.

Besides, it is [recommended](https://news.ycombinator.com/item?id=26776956)
to write `if [ "x$ENABLE_MORPHOLOGY_ANALYSIS" = "xtrue" ]; `

Best regards,
Maoyin

**Comment by Crissium** [comment]
Thanks for your explanation! I have updated the dockerfile to use Python packages from the Alpine repo if present (for maximum compatibility). Later I'll look into how to set up CI to build and push the image automatically.

Thank you again for your contribution! I believe docker should be the recommended way to run SilverDict on Mac for technologically unsophisticated users.

**Comment by simonmysun** [comment]
I would like to remind that I anchored the base image versions to ensure reproducibility. For security reasons, it needs to be kept up-to-date, like other upstream libraries.

**Comment by Crissium** [comment]
> I would like to remind that I anchored the base image versions to ensure reproducibility. For security reasons, it needs to be kept up-to-date, like other upstream libraries.

Thanks for the reminder! Which do you think is better, like this:

```dockerfile
# Stage 1: Frontend Builder
FROM node:lts-alpine as frontend-builder

# ...

# Stage 2: Production Environment
FROM alpine
```

Or just update the base images from time to time?

**Comment by simonmysun** [comment]
I prefer a fixed version for reproducibility, otherwise users may encounter issues that aren't apparent in your environment. Updating becomes imperative when the base image is identified as vulnerable. For example both images here are affected by CVE-2024-2511 (ref: https://hub.docker.com/layers/library/alpine/3/images/sha256-15c46ced65c6abed6a27472a7904b04273e9a8091a5627badd6ff016ab073171).

This indeed introduces extra work. A compromise approach is to fix only the major and minor versions, leaving the patch version and hash unfixed. This helps minimize inconsistency issues while still allowing users to receive security patches.

**Comment by arabshapt** [comment]
hey guys! did you solve the problem with xapian? Using https://hub.docker.com/r/mathdodger/silverdict I am still not able to do full text search... I get this error: Xapian group does not exist. Unable to proceed.
My version of the app is 1.2.2

**Comment by Crissium** [comment]
> hey guys! did you solve the problem with xapian? Using https://hub.docker.com/r/mathdodger/silverdict I am still not able to do full text search... I get this error: Xapian group does not exist. Unable to proceed.

Please take a look at https://github.com/Crissium/SilverDict/wiki/general-notes#full-text-search

Now groups are used to mark dictionaries that should be indexed for full-text search, loaded into memories, etc. Perhaps this is not intuitive or user-friendly and should be changed.

**Comment by arabshapt** [comment]
Thank you for quick and helpful response:)

**Pull request merged**